File: src/year2025/day1/day1.py
#!/usr/bin/env python3
# [Day 1: Secret Entrance](https://adventofcode.com/2025/day/1)

import atexit
import time
from argparse import ArgumentParser
from pathlib import Path

parser = ArgumentParser()
parser.add_argument("-v", "--verbose", action="store_true")
parser.add_argument("-t", "--test", action="store_true")
parser.add_argument("--elapsed", action="store_true")
parser.add_argument("filename", nargs="?", type=Path, default="input.txt")
args = parser.parse_args()
if args.test:
    args.filename = Path("test.txt")

data = args.filename.read_text()

if args.elapsed:
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


def part1(ops):
    pos = 50
    count_zero = 0

    for d, n in ops:
        if d == "L":
            pos = (pos - n) % 100
        else:  # R
            pos = (pos + n) % 100

        if pos == 0:
            count_zero += 1

    return count_zero


def part2(ops):
    pos = 50
    count_zero = 0

    for d, n in ops:
        step = -1 if d == "L" else 1

        for _ in range(n):
            pos = (pos + step) % 100
            if pos == 0:
                count_zero += 1

    return count_zero


ops = [(line[0], int(line[1:])) for line in data.splitlines()]
print(part1(ops))
print(part2(ops))

File: src/year2025/day1/day1.go
// [Day 1: Secret Entrance](https://adventofcode.com/2025/day/1)

package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
	"time"
)

type Op struct {
	dir byte
	num int
}

// part1 computes the password to open the door.
func part1(ops []Op) int {

	pos := 50

	countZero := 0

	for _, op := range ops {

		if op.dir == 'L' {
			pos = (pos - op.num + 100) % 100
		} else { // 'R'
			pos = (pos + op.num) % 100
		}

		if pos == 0 {
			countZero++
		}
	}

	return countZero
}

// part2 computes the password to open the door according to password method 0x434C49434B (aka 'CLCK').
func part2(ops []Op) int {

	pos := 50

	countZero := 0

	for _, op := range ops {

		step := 1 // 'R'
		if op.dir == 'L' {
			step = -1
		}

		for i := 0; i < op.num; i++ {
			pos = (pos + step + 100) % 100
			if pos == 0 {
				countZero++
			}
		}
	}

	return countZero
}

// solve solves parts one and two of the puzzle.
func solve(data []byte) (int, int) {

	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	ops := make([]Op, 0, len(lines))

	for _, line := range lines {
		num, err := strconv.Atoi(line[1:])
		if err == nil {
			ops = append(ops, Op{dir: line[0], num: num})
		}

	}

	return part1(ops), part2(ops)
}

func main() {
	filename := "input.txt"
	elapsed := false

	for i := 1; i < len(os.Args); i++ {
		arg := os.Args[i]
		if arg == "--elapsed" {
			elapsed = true
		} else if !strings.HasPrefix(arg, "-") {
			filename = arg
		}
	}

	data, err := os.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}

	if elapsed {
		start := time.Now()
		result1, result2 := solve(data)
		duration := time.Since(start)

		fmt.Println(result1)
		fmt.Println(result2)
		fmt.Printf("elapsed: %f ms\n", duration.Seconds()*1000.0)
	} else {
		result1, result2 := solve(data)

		fmt.Println(result1)
		fmt.Println(result2)
	}
}
